sim/core/rune.go

```diff
@@ -0,0 +1,208 @@
+package core
+
+import (
+	"time"
+)
+
+type OnRuneGain func(sim *Simulation)
+
+const (
+	RuneType_Blood int32 = iota
+	RuneType_Frost
+	RuneType_Unholy
+	RuneType_Death
+)
+
+const (
+	RuneTypeBloodName  string = "Blood"
+	RuneTypeFrostName  string = "Frost"
+	RuneTypeUnholyName string = "Unholy"
+	RuneTypeDeathName  string = "Death"
+	RuneTypeUndefName  string = "Undefined"
+)
+
+func (rb *runeBar) RuneTypeName() string {
+	switch rb.runeType {
+	case RuneType_Blood:
+		return RuneTypeBloodName
+	case RuneType_Frost:
+		return RuneTypeFrostName
+	case RuneType_Unholy:
+		return RuneTypeUnholyName
+	case RuneType_Death:
+		return RuneTypeDeathName
+	}
+	// Should never get this!
+	return RuneTypeUndefName
+}
+
+type AvailableRune struct {
+	slot      int32
+	available bool
+}
+
+type runeBar struct {
+	unit *Unit
+
+	runeType int32
+
+	maxRunes     float64
+	currentRunes float64
+
+	cooldowns [3]Cooldown
+
+	onRuneGain OnRuneGain
+}
+
+func (unit *Unit) EnableRuneBars(onBloodRuneGain OnRuneGain,
+	onFrostRuneGain OnRuneGain,
+	onUnholyRuneGain OnRuneGain,
+	onDeathRuneGain OnRuneGain) {
+
+	unit.runicPowerBar.bloodRunesBar = runeBar{
+		unit: unit,
+
+		runeType: RuneType_Blood,
+
+		maxRunes:     2,
+		currentRunes: 2,
+
+		cooldowns: [3]Cooldown{
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{nil, 0},
+		},
+
+		onRuneGain: onBloodRuneGain,
+	}
+
+	unit.runicPowerBar.frostRunesBar = runeBar{
+		unit: unit,
+
+		runeType: RuneType_Frost,
+
+		maxRunes:     2,
+		currentRunes: 2,
+
+		cooldowns: [3]Cooldown{
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{nil, 0},
+		},
+
+		onRuneGain: onFrostRuneGain,
+	}
+
+	unit.runicPowerBar.unholyRunesBar = runeBar{
+		unit: unit,
+
+		runeType: RuneType_Unholy,
+
+		maxRunes:     2,
+		currentRunes: 2,
+
+		cooldowns: [3]Cooldown{
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{nil, 0},
+		},
+
+		onRuneGain: onUnholyRuneGain,
+	}
+
+	unit.runicPowerBar.deathRunesBar = runeBar{
+		unit: unit,
+
+		runeType: RuneType_Death,
+
+		maxRunes:     3,
+		currentRunes: 0,
+
+		cooldowns: [3]Cooldown{
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+			Cooldown{unit.NewTimer(), 10 * time.Second},
+		},
+
+		onRuneGain: onDeathRuneGain,
+	}
+}
+
+func (unit *Unit) HasBloodRuneBar() bool {
```

Comment:
It should be fine to have only 1 of these functions (i.e. HasRunicPower) since they're all connected.